lib/VolSurface/Calibration/SABR.pm

```diff
@@ -0,0 +1,514 @@
+package VolSurface::Calibration::SABR;
+
+=head1 NAME
+
+VolSurface::Calibration::SABR
+
+=head1 DESCRIPTION
+
+SABR is a calibration model that is used in volatility surface calibration.
+
+=cut
+
+use Moose::Role;
+
+requires qw(symbol surface term_by_day smile_points parameterization);
+
+use Date::Utility;
+use List::MoreUtils qw(pairwise first_index indexes);
+use List::Util qw(min max);
+use Math::Trig qw(tanh);
+use Try::Tiny;
+use Format::Util::Numbers qw(roundnear);
+
+our $VERSION = '0.01';
+
+has _max_iteration => (
+    is      => 'ro',
+    isa     => 'Int',
+    default => 500,
+);
+
+has _tolerance_level => (
+    is      => 'ro',
+    isa     => 'Num',
+    default => 1e-05,
+);
+
+has _scale => (
+    is      => 'ro',
+    isa     => 'Num',
+    default => 10,
+);
+
+=head2 function_to_optimize
+The function that we want to optimize.
+=cut
+
+sub function_to_optimize {
+    my ($self, $params) = @_;
+
+    # this number is returned when the optimization is heading the wrong path
+    my $error = 1000;
+
+    #if (not $params) {
+    #    $self->_logger->logcroak('initial params are not present during optimization');
+    #}
+
+    my %params =
+        pairwise { our $a => our $b } @{$self->calibration_param_names},
+        @$params;
+
+    my @tenors = @{$self->term_by_day};
+    my @kurtosis =
+        map { $self->_calculate_kurtosis($_ / 365, \%params) } @tenors;
+    my @variance =
+        map { $self->_calculate_variance($_ / 365, \%params) } @tenors;
+    my @skew = map { $self->_calculate_skew($_ / 365, \%params) } @tenors;
+
+    foreach my $param_name (keys %params) {
```

Comment:
There is no order in the output of `keys`. To get a stable result it's possibly better to use `@{$self->calibration_param_names}` here.